cmd/bosun/expr/annotate.go

```diff
@@ -0,0 +1,254 @@
+package expr
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"math"
+
+	"bosun.org/cmd/bosun/expr/parse"
+	"bosun.org/models"
+	"bosun.org/opentsdb"
+	"github.com/MiniProfiler/go/miniprofiler"
+	"github.com/bosun-monitor/annotate"
+	"github.com/kylebrandt/boolq"
+)
+
+var Annotate = map[string]parse.Func{
+	// Funcs for querying elastic
+	"ancounts": {
+		Args:   []models.FuncType{models.TypeString, models.TypeString, models.TypeString},
+		Return: models.TypeSeriesSet,
+		Tags:   tagFirst,
+		F:      AnCounts,
+	},
+	"andurations": {
+		Args:   []models.FuncType{models.TypeString, models.TypeString, models.TypeString},
+		Return: models.TypeSeriesSet,
+		Tags:   tagFirst,
+		F:      AnDurations,
+	},
+	"antable": {
+		Args:   []models.FuncType{models.TypeString, models.TypeString, models.TypeString, models.TypeString},
+		Return: models.TypeTable,
+		F:      AnTable,
+	},
+}
+
+func procDuration(e *State, startDuration, endDuration string) (*time.Time, *time.Time, error) {
+	start, err := opentsdb.ParseDuration(startDuration)
+	if err != nil {
+		return nil, nil, err
+	}
+	var end opentsdb.Duration
+	if endDuration != "" {
+		end, err = opentsdb.ParseDuration(endDuration)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+	st := e.now.Add(time.Duration(-start))
+	en := e.now.Add(time.Duration(-end))
+	return &st, &en, nil
+}
+
+func getAndFilterAnnotations(e *State, start, end *time.Time, filter string) (annotate.Annotations, error) {
+	annotations, err := e.AnnotateContext.GetAnnotations(start, end, "", "", "", "", "", "", "")
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations := annotate.Annotations{}
+	for _, a := range annotations {
+		if filter == "" {
+			filteredAnnotations = append(filteredAnnotations, a)
+			continue
+		}
+		match, err := boolq.AskExpr(filter, a)
+		if err != nil {
+			return nil, err
+		}
+		if match {
+			filteredAnnotations = append(filteredAnnotations, a)
+		}
+	}
+	sort.Sort(sort.Reverse(annotate.AnnotationsByStartID(filteredAnnotations)))
+	return filteredAnnotations, nil
+}
+
+func AnDurations(e *State, T miniprofiler.Timer, filter, startDuration, endDuration string) (r *Results, err error) {
+	reqStart, reqEnd, err := procDuration(e, startDuration, endDuration)
+	if err != nil {
+		return nil, err
+	}
+	filteredAnnotations, err := getAndFilterAnnotations(e, reqStart, reqEnd, filter)
+	if err != nil {
+		return nil, err
+	}
+	series := make(Series)
+	for i, a := range filteredAnnotations {
+		aStart := a.StartDate.Time
+		aEnd := a.EndDate.Time
+		inBounds := (aStart.After(*reqStart) || aStart == *reqStart) && (aEnd.Before(*reqEnd) || aEnd == *reqEnd)
+		entirelyOutOfBounds := aStart.Before(*reqStart) && aEnd.After(*reqEnd)
+		aDuration := aEnd.Sub(aStart)
+		if inBounds {
+			// time has no meaning here, so we just make the key an index since we don't have an array type
+			series[time.Unix(int64(i), 0).UTC()] = aDuration.Seconds()
```

Comment:
I hate this misuse of timestamps. It is pretty much useless without some aggregation (I'm assuming sum) is applied on top. So why not aggregate here and return a number?